### PR TITLE
videoio: plugins API versioning

### PR DESCRIPTION
**Pull request by alalek**
Revised support for plugin API versioning.

**RFC**

/cc @mshabunin @asmorkalov @joshdoe

Thanks for looking at these and providing feedback.

<cut/>

To be merged after the current release.

### CODE REVIEW

File: modules/videoio/src/backend_plugin.cpp [hunk 1]
```diff
@@ -534,23 +534,23 @@ class PluginWriter : public cv::IVideoWriter
     {
         CV_Assert(plugin_api);
         CvPluginWriter writer = NULL;
-        if (plugin_api->api_header.api_version >= 1 && plugin_api->Writer_open_with_params)
+        if (plugin_api->api_header.api_version >= 1 && plugin_api->v1.Writer_open_with_params)
```
Comment (alalek) [inline modules/videoio/src/backend_plugin.cpp:537]: `v0` access doesn't require `api_version` checks
`v1`+ access requires `api_version` check and compatibility warnings about missing features `CV_LOG_WARNING(... due to lower API level)`

File: modules/videoio/src/cap_gstreamer.cpp [hunk 2]
```diff
@@ -1839,6 +1839,8 @@ void handleMessage(GstElement * pipeline)
 
 #if defined(BUILD_PLUGIN)
 
+#define ABI_VERSION 0
+#define API_VERSION 1
 #include "plugin_api.hpp"
```
Comment (alalek) [inline modules/videoio/src/cap_gstreamer.cpp:1844]: `ABI_VERSION` / `API_VERSION` must be defined on plugins size (`defined(BUILD_PLUGIN)`) before including `plugin_api.hpp`

**Review by alalek — COMMENTED** [review]
It make sense to split on 2 interfaces for VideoCapture and VideoWriter

**Review by mshabunin — APPROVED** [review]

**Comment by mshabunin** [comment]
Looks like we forgot to add ABI/API macros to the mfx plugin
```.cpp
#define ABI_VERSION 0
#define API_VERSION 0
```

**Comment by alalek** [comment]
Please check fix here: #19321

**Pull request merged**